clients/unshaded/src/main/java/tachyon/client/TachyonFile.java

```diff
@@ -198,17 +202,24 @@ public String getLocalFilename(int blockIndex) throws IOException {
    * @throws IOException if the underlying file does not exist or its metadata is corrupted
    */
   public List<String> getLocationHosts() throws IOException {
-    List<String> ret = new ArrayList<String>();
+    Set<String> ret = Sets.newHashSet();
     if (getNumberOfBlocks() > 0) {
-      List<NetAddress> locations = getClientBlockInfo(0).getLocations();
-      if (locations != null) {
-        for (NetAddress location : locations) {
+      // under FS locations
+      List<NetAddress> underFsLocations = getClientBlockInfo(0).getUnderFsLocations();
+      if (underFsLocations != null) {
+        for (NetAddress location : underFsLocations) {
           ret.add(location.host);
         }
       }
+      List<BlockLocation> blockLocations = getClientBlockInfo(0).getBlockInfo().getLocations();
```

Comment:
Done. Note that the returned list may contain duplicates, but it already has this issue in the old behavior.